[ML] Integrates into serverless kibana

jgowdyelastic opened the pull request:
- Adds function to ML plugin server side setup contract (`setFeaturesEnabled`) to enable/disable features. Features being, anomaly detection, data frame analytics or natural language processing.
  - Each serverless plugin (search, observability and security) now calls `setFeaturesEnabled`  to enable different features on setup.
- Adds three new feature capabilities, `isADEnabled`, `isDFAEnabled`, `isNLPEnabled` 
  - Updates the capabilities switcher to toggle these capabilities based on the shared feature API. This currently has a bug where the switcher isn't triggered for API tag checking.
- deeplinks are now only registered based on the feature capabilities, these control what is shown in the nav menu for search and observability projects.
- Adds the Machine Learning nav menu section to the `serverless_search` plugin, to match the `serverless_observability` plugin.
- Does not updated the `serverless_security` plugin's nav menu. (update, `serverless_observability` nav menu also now does not contain ML)

Relates to https://github.com/elastic/kibana/issues/160891

jgowdyelastic committed:
[ML] Integrating into serverless kibana

jgowdyelastic committed:
removing temp debugging code

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

jgowdyelastic committed:
adding type keyword

jgowdyelastic committed:
Merge remote-tracking branch 'origin/main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge remote-tracking branch 'origin/main' into ml-serverless-compatibility

jgowdyelastic committed:
updating navigation menus for es and observability

jgowdyelastic committed:
updating capabilities switcher

jgowdyelastic committed:
correcting icon

jgowdyelastic committed:
Merge remote-tracking branch 'origin/main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge remote-tracking branch 'origin/main' into ml-serverless-compatibility

jgowdyelastic committed:
registering deeplinks based on features

jgowdyelastic committed:
fixing types

jgowdyelastic committed:
disabling notifications

jgowdyelastic committed:
fixing types

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
fixing mocked nav

jgowdyelastic committed:
fixing capabilities tests

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge remote-tracking branch 'origin/main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
filtering notifications based on type

jgowdyelastic committed:
adding missing import

jgowdyelastic commented on x-pack/plugins/serverless_search/kibana.jsonc:
```diff
@@ -9,18 +9,19 @@
-      "devTools",
       "console",
-      "searchprofiler",
+      "dashboard",
```
adding `ml` but also sorting the list alphabetically

jgowdyelastic committed:
removing comment

jgowdyelastic committed:
updating comment

jgowdyelastic committed:
type keyword

jgowdyelastic committed:
page access checks

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

sphilipse reviewed (approved):
Serverless-search changes LGTM

darnautov commented on x-pack/plugins/ml/common/types/capabilities.ts:
```diff
@@ -182,3 +198,50 @@ export interface MlCapabilitiesResponse {
 export type ResolveMlCapabilities = (request: KibanaRequest) => Promise<MlCapabilities | null>;
+
+interface FeatureCapabilities {
+  ad: MlCapabilitiesKey[];
+  dfa: MlCapabilitiesKey[];
+  nlp: MlCapabilitiesKey[];
```
should it be different capabilities for `ad`, `dfa` and `npl`?  
You can extract the type from the `featureCapabilities` constant below.

jgowdyelastic replied:
The interface here is only used for `featureCapabilities` below, it is not exported out. So it's a bit of a chicken an egg situation.

darnautov commented on x-pack/plugins/ml/public/application/components/ml_page/ml_page.tsx:
```diff
@@ -124,13 +126,19 @@ export const MlPage: FC<{ pageDeps: PageDependencies }> = React.memo(({ pageDeps
-          items: useSideNavItems(activeRoute),
-        }}
+        // forcing the background to white navigation is disabled
+        css={pageDeps.navMenuEnabled ? {} : { background: '#FFF' }}
```
I guess we should use the EUI variable instead

darnautov commented on x-pack/plugins/ml/public/plugin.ts:
```diff
@@ -99,6 +100,8 @@ export class MlPlugin implements Plugin<MlPluginSetup, MlPluginStart> {
 
   private locator: undefined | MlLocator;
 
+  private navMenuEnabled = true;
```
This value has to be an Observable. At the moment there is no way the app will get an update if someone calls `setNavMenuEnabled.

jgowdyelastic replied:
This value is set on kibana setup and does not change during the running on kibana as kibana will not switch from being in serverless to not being in serverless.

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/159433/commits/8f4e744df95954be50b562cbb3fff61a386d46d5

darnautov commented on x-pack/plugins/ml/public/register_helper/register_search_links/search_deep_links.ts:
```diff
@@ -8,203 +8,286 @@
+    }),
+    path: `/${ML_PAGES.OVERVIEW}`,
+    navLinkStatus:
+      mlCapabilities.isADEnabled && mlCapabilities.isDFAEnabled && mlCapabilities.isNLPEnabled
```
I wonder if we should require all these capabilities for the Overview page? Perhaps we can hide some panels if certain capabilities are missing?

jgowdyelastic replied:
The overview page is not currently available in serverless.
Having all three of these feature capabilities required is essentially like saying `isServerless === false`.
This may change if one of the projects decides they want the overview page added, in which case, yes we will need to hide/show panels based on these feature capabilities.

qn895 replied:
Would be good to add a comment or have a util function for now for better clarity if that's the case.

jgowdyelastic replied:
updated in https://github.com/elastic/kibana/pull/159433/commits/c29461ae8a78500890a6fb0ca638aa51a449d80b

jgowdyelastic replied:
It is possible to access the overview page from the breadcrumbs.
Because of this I've [adjusted](https://github.com/elastic/kibana/pull/159433/commits/82a51a71339a83a1785a45536dba3b62cc41e8e0) the overview page to hide panels based on feature enablement. 
I've added note the the [meta issue](https://github.com/elastic/kibana/issues/160891) for this this task to say that further decisions will be needed by the various serverless plugins as to whether they want an ML overview page included. If they decided they do they can add a nav menu item in and we can adjust these deep links checks.

For now I'm going to keep the overview page out of the nav menu because only the Search plugin will see it, and for that plugin in the overview page is completely blank as it has AD and DFA turned off.

darnautov commented on x-pack/plugins/ml/public/register_helper/register_search_links/search_deep_links.ts:
```diff
@@ -8,203 +8,286 @@
+    }),
+    path: `/${ML_PAGES.DATA_VISUALIZER}`,
+    navLinkStatus:
+      mlCapabilities.isADEnabled && mlCapabilities.isDFAEnabled && mlCapabilities.isNLPEnabled
```
do we need all of these for the data visualizer?

jgowdyelastic replied:
For now, both data visualizer pages are hidden in serverless ML. This may change in the future if one of the projects decides they want to include them in their nav menu.

darnautov commented on x-pack/plugins/ml/public/register_helper/register_search_links/search_deep_links.ts:
```diff
@@ -8,203 +8,286 @@
+    keywords: ['CSV', 'JSON'],
+    path: `/${ML_PAGES.DATA_VISUALIZER_FILE}`,
+    navLinkStatus:
+      mlCapabilities.isADEnabled && mlCapabilities.isDFAEnabled && mlCapabilities.isNLPEnabled
```
does the file upload require all of these?

darnautov commented on x-pack/plugins/ml/public/register_helper/register_search_links/search_deep_links.ts:
```diff
@@ -8,203 +8,286 @@
+    }),
+    path: `/${ML_PAGES.DATA_VISUALIZER_INDEX_SELECT}`,
+    navLinkStatus:
+      mlCapabilities.isADEnabled && mlCapabilities.isDFAEnabled && mlCapabilities.isNLPEnabled
```
does the index visualizer require all of these?

darnautov commented on x-pack/plugins/ml/public/register_helper/register_search_links/search_deep_links.ts:
```diff
@@ -8,203 +8,286 @@
+      defaultMessage: 'Notifications',
+    }),
+    path: `/${ML_PAGES.NOTIFICATIONS}`,
+    navLinkStatus: AppNavLinkStatus.visible,
```
shall we add an appropriate capabilities check for the notifications page?

jgowdyelastic replied:
The notification page is always available in serverless and non-serverless modes.

darnautov commented on x-pack/plugins/ml/server/plugin.ts:
```diff
@@ -93,6 +99,11 @@ export class MlServerPlugin
   private isMlReady: Promise<void>;
   private setMlReady: () => void = () => {};
   private savedObjectsSyncService: SavedObjectsSyncService;
+  private enabledFeatures: MlFeatures = {
```
would it be possible to disable/enable some features during runtime? perhaps this value should be an Observable in that case.

jgowdyelastic replied:
It is possible to switch serverless project while kibana is running, when this happens the kibana server performs a hard restart so it can initialise a different serverless plugin. It is during this serverless plugin setup that our `setFeaturesEnabled` function is called setting these `enabledFeatures` flags.

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
removing nav menu shared function

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

jgowdyelastic committed:
removing old isServerless function

jgowdyelastic committed:
removing old isServerless hook

jgowdyelastic committed:
fixing background color

jgowdyelastic committed:
removing background color overrides

jgowdyelastic committed:
fixing mock

kpatticha reviewed (changes requested):
it seems the deeps links are not registered in oblt 
<img width="1792" alt="image" src="https://github.com/elastic/kibana/assets/3369346/ff184d8b-31c7-4e7a-9f36-2a8ab4f7c749">

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic commented:
@kpatticha That's odd. It works ok for me. This is the nav menu I see.
<img width="231" alt="image" src="https://github.com/elastic/kibana/assets/22172091/9097273c-03f0-46b6-9d5b-32d64c759657">

How are you running kibana?
I'm using `yarn serverless-oblt`

kpatticha commented:
> I suspect ML just isn't enabled at all in the es that you are running

hmm, that could be. Are you running --trial ?

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
fixing links when not in serverless

jgowdyelastic commented:
> On non-serverless mode, the navigation is now like this. Is this intentional?

No, not intentional, this has been fixed [here](https://github.com/elastic/kibana/pull/159433/commits/c29461ae8a78500890a6fb0ca638aa51a449d80b)

jgowdyelastic committed:
adding ml back into oblt menu

jgowdyelastic commented:
@kpatticha Apologies, it looks like you were right, the ML menu had very recently been removed from the observability menu in this [PR](https://github.com/elastic/kibana/pull/160620) 
I had merged this PR with main via github, but not updated my local copy, hence me still seeing it in the menu.

I have updated this PR to put the ML menu back in. Once this PR is in, I believe the menus will be the responsibility of the serverless plugin teams, so the ML links can be moved about wherever needed. I've just added an ML section in this PR as an initial step.

kpatticha commented on x-pack/plugins/serverless_observability/public/components/side_navigation/index.tsx:
```diff
@@ -94,6 +94,10 @@ const navigationTree: NavigationTreeDefinition = {
       ],
     },
+    {
+      type: 'navGroup',
```
In the recent mockups, we won't use the ml presets for the observability side nav but instead a custom node tree. It would be great if you revert this change.

jgowdyelastic replied:
I've removed this ML section https://github.com/elastic/kibana/pull/159433/commits/3a7eae13415e0a6f36583eacf1b6a6929038baca

It's worth pointing out for other reviews that until these custom ML links are added, there will be no way to access ML pages from the side navigation menu when running the Observability project.
It will still be possible to access ML pages from kibana's app search feature at the top of kibana.

kpatticha reviewed (approved):
oblt LGTM. I just left a small comment regarding the preset.

jgowdyelastic committed:
removing ml section from oblt

jgowdyelastic committed:
adjusting overview page

jgowdyelastic committed:
fixing canViewMlNodes value for viewer users

loixlab reviewed (approved):
LGTM 👍

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

jgowdyelastic committed:
Merge branch 'main' into ml-serverless-compatibility

qn895 commented:
Tested latest changes and LGTM 🎉
